src/azul/queues.py

```diff
@@ -149,66 +150,98 @@ def get_queues(self, queue_names: Iterable[str]) -> Mapping[str, Queue]:
             queue_length = sum(message_counts)
             logger.debug('Queue %s has %i message(s) (%i available, %i in flight and %i delayed).',
                          queue_name, queue_length, *message_counts)
-            total_message_count += queue_length
-        return total_message_count
-
-    def wait_for_queue_level(self, empty: bool = True, num_bundles: int = None):
+            queue_message_count[queue_name] = queue_length
+            queue_message_count['total_message_count'] += queue_length
+        return queue_message_count
+
+    def wait_for_queue_level(self,
+                             empty: bool = True,
+                             num_bundles: int = None,
+                             min_timeout: Optional[int] = None,
+                             max_timeout: Optional[int] = None):
         """
         Wait until the total count of messages in the notify and document queues
         reaches the desired level
 
         :param empty: True to wait until the queues are empty, False to wait until not empty.
         :param num_bundles: Number of bundles being indexed (None = many bundles)
+        :param min_timeout: Specifies minimum wait time for queues to process
+                            index notifications.
+        :param max_timeout: Specifies maximum wait time for queues to process
+                            index notifications.
         """
+
+        def _compute_timeout(num_bundles):
+            # It takes approx. 6 seconds per worker to process a bundle.
+            timeout = 6 * num_bundles / config.indexer_concurrency
+            if max_timeout is not None:
+                logger.info('Maximum wait time set to %s', max_timeout)
+                timeout = min(timeout, max_timeout)
+            if min_timeout is not None:
+                logger.info('Minimum wait time set to %s', max_timeout)
+                timeout = max(min_timeout, timeout)
+            return timeout
+
         sleep_time = 5
         deque_size = 10 if empty else 1
         queues = self.get_queues(config.work_queue_names)
         queue_size_history = deque(maxlen=deque_size)
         wait_start_time = time.time()
+        dynamic_timeout = False
 
         if not empty:
             timeout = 2 * 60
-        elif num_bundles is None:
-            timeout = 60 * 60
         else:
-            # calculate timeout for queues to empty with a given number of bundles
-            time_to_clear_deque = deque_size * sleep_time
-            time_processing_bundles = num_bundles * 5  # small time per bundle
-            timeout = max(time_processing_bundles + time_to_clear_deque, 60)  # at least 1 min
-            timeout = min(timeout, 60 * 60)  # at most 60 min
+            if num_bundles is None:
+                queue_message_count = self.count_messages(queues)
+                num_bundles = queue_message_count[config.notify_queue_name]
```

Comment:
Other than for the log statement in line 195, timeout isn't really used until it is reassigned. I would like you to try to eliminate the `dynamic_timeout` variable and instead use `timeout is None` to a signal dynamic timeout and you should also try to roll the loop differently so that the log statement that logs the updated timeout and the one that logs the initial timeout can be replaced with a single log statement at the beginning of the loop body.